drivers/input/input_pmw3610.c

```diff
@@ -0,0 +1,433 @@
+	}
+
+	k_sleep(K_MSEC(INIT_OBSERVATION_DELAY_MS));
+
+	ret = pmw3610_read_reg(dev, PMW3610_OBSERVATION1, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((val & OBSERVATION1_INIT_MASK) != OBSERVATION1_INIT_MASK) {
+		LOG_ERR("Unexpected OBSERVATION1 value: %02x", val);
+		return -EINVAL;
+	}
+
+	for (uint8_t reg = PMW3610_MOTION; reg <= PMW3610_DELTA_XY_H; reg++) {
+		ret = pmw3610_read_reg(dev, reg, &val);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = pmw3610_write_reg(dev, PMW3610_PERFORMANCE, PERFORMANCE_INIT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = pmw3610_write_reg(dev, PMW3610_RUN_DOWNSHIFT, RUN_DOWNSHIFT_INIT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = pmw3610_write_reg(dev, PMW3610_REST1_RATE, REST1_RATE_INIT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = pmw3610_write_reg(dev, PMW3610_REST1_DOWNSHIFT, REST1_DOWNSHIFT_INIT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Configuration */
+
+	if (cfg->invert_x || cfg->invert_y) {
+		ret = pmw3610_spi_clk_on(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = pmw3610_write_reg(dev, PWM3610_SPI_PAGE0, SPI_PAGE0_1);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = pmw3610_read_reg(dev, PMW3610_RES_STEP, &val);
+		if (ret < 0) {
+			return ret;
+		}
+
+		WRITE_BIT(val, RES_STEP_INV_X_BIT, cfg->invert_x);
+		WRITE_BIT(val, RES_STEP_INV_Y_BIT, cfg->invert_y);
+
+		ret = pmw3610_write_reg(dev, PMW3610_RES_STEP, val);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = pmw3610_write_reg(dev, PWM3610_SPI_PAGE1, SPI_PAGE1_0);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = pmw3610_spi_clk_off(dev);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int pmw3610_init(const struct device *dev)
+{
+	const struct pmw3610_config *cfg = dev->config;
+	struct pmw3610_data *data = dev->data;
+	int ret;
+
+	if (!spi_is_ready_dt(&cfg->spi)) {
+		LOG_ERR("%s is not ready", cfg->spi.bus->name);
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init(&data->motion_work, pmw3610_motion_work_handler);
+
+	if (!gpio_is_ready_dt(&cfg->motion_gpio)) {
+		LOG_ERR("%s is not ready", cfg->motion_gpio.port->name);
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&cfg->motion_gpio, GPIO_INPUT);
+	if (ret != 0) {
+		LOG_ERR("Motion pin configuration failed: %d", ret);
+		return ret;
+	}
+
+	gpio_init_callback(&data->motion_cb, pmw3610_motion_handler,
+			   BIT(cfg->motion_gpio.pin));
+
+	ret = gpio_add_callback_dt(&cfg->motion_gpio, &data->motion_cb);
+	if (ret < 0) {
+		LOG_ERR("Could not set motion callback: %d", ret);
+		return ret;
+	}
+
+	ret = pmw3610_configure(dev);
+	if (ret != 0) {
+		LOG_ERR("Device configuration failed: %d", ret);
+		return ret;
+	}
+
+	ret = gpio_pin_interrupt_configure_dt(&cfg->motion_gpio,
+					      GPIO_INT_EDGE_TO_ACTIVE);
+	if (ret != 0) {
+		LOG_ERR("Motion interrupt configuration failed: %d", ret);
+		return ret;
+	}
+
+	ret = pm_device_runtime_enable(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable runtime power management");
```

Comment:
it's alwaus useful to log `ret` value

Reply:
cool, added, gotta fix in in few other drivers too, will followup